rfcs: parser as a go sub-module (#28015)
pingcap/tidb: docs/design/2021-09-13-parser-as-submodule-of-tidb.md

# Proposal: Move parser back to pingcap/tidb

- Author(s): [xhebox](http://github.com/xhebox)
- Discussion: [Move parser back to pingcap/tidb](https://internals.tidb.io/t/topic/385)

> - Tracking Issue: None for now.

Reviewer: @xhebox please create an tracking issue before merging this RFC.

Author: I won't merge it now, since I've promised to have 3 LGTM and wait for 3 days. I would create one later, you could give the 3rd LGTM.

Reviewer: OK. I'll review this proposal today.

Author: #28257, here it is.

> # TiDB Design Documents

Reviewer 2: ```suggestion
# Proposal: Move parser back to pingcap/tidb
```

2. It is a mysql parser, but is closely related to TiDB(basically everything except ast/mysql packages). You must code TiDB logic into the "standalone mysql parser" first before your real implementation, which means at least two PRs, even for small changes.

3. Before the PR of parser merged, the PR of TiDB must use module directive to redirect a specific version of parser to fix CI.

Problems above have defeated the previous argument `only go.mod need to change accordingly`. We only separated code into two repositories, actual work did not decrease. However, our development experience does getting worse.

## Detailed Plan

Obviously, we could move back to the old good way. But it must be a go sub-module, because:

1. To keep compatibility. Parser has been imported by other tools and other users.
2. To keep the clean dependency. That is why it is separated. We don't want to make things worse again.

The detailed plan:

1. Create parser directory, a sub-module, in pingcap/tidb, which is synchronized with pingcap/parser, with `replace github.com/pingcap/parser => ./parser`.

> 2. Stop sending new PRs and issues to pingcap/parser, new PRs should be directly sent to pingcap/tidb. And merge recent PRs in pingcap/parser as much as possible. Changes will be synchronized from pingcap/parser to pingcap/tidb(manually), of course.

Reviewer: ```suggestion
2. Stop sending new PRs and issues to pingcap/parser, new PRs should be directly sent to pingcap/tidb. And merge recent PRs in pingcap/parser as much as possible. Changes will be synchronized from pingcap/parser to pingcap/tidb manually, of course.
```

> 3. Announce the deprecation of pingcap/parser to possible users. Clean up old issues and PRs, create it in pingcap/tidb again, close it, or just forget about it.

Reviewer 3: (issues can be transferred, no need to close and reopen)

Author: https://docs.github.com/en/issues/tracking-your-work-with-issues/transferring-an-issue-to-another-repository, this, right? I know...

Author: Updated.

4. We can create wrapper packages for pingcap/parser, which re-exports things from `pingcap/tidb/parser`. I mean something like `import ( . github.com/pingcap/tidb/parser/xxx)`, create a dummy function will eliminate the error of `not used import`. All code of pingcap/parser will be removed in this step. This will delevery new updates without the need of migrating importing paths.
5. After another one or two dev cycles, we could archive pingcap/parser. It mainly depends on users of pingcap/parser.

> 

Reviewer: You should also mention how to migrate the test infra.

Reviewer: OK I see it below, but please consider it in the implementation plan. It is ended by migrated, not only answered the question.

Author: Added in step 3.

There will be a tracking issue, and the whole progress is public. Step 1 or 2 will likely take one or two weeks. Internally we could do the migration for internal tools while doing step 3.

Step 4 and 5 are long-running tasks to smoothly end the support of the old parser.

## Questions

- What about projects importing parser?

Nothing needs to be done before step 5. However, one will need to migrate the import path eventually, before the final deprecation.

- What about some complex dependency chain? For example, `pkg -> forked tidb -> other pingcap/parser`.

Nothing different for the complex dependencies. Just do the migration before the deprecation. And we don't actually remove the old commits for not breaking old apps.

- What about CI and tests?

It is not a problem. pingcap/parser only has unit tests, and integration tests `make gotest` of `pingcap/tidb`. We could run integration tests of tidb, and only run unit tests.

- How about release and versioning?

> Since sub-module is merely another module inside another module, we need to release it separately from the main module. That means our release team needs to tag `parser/vX.X.X`, which can be solved by a script: thanks to the identical release cycle/version between parser and tidb.

Reviewer 3: or we could use true semver for the parser sub-module instead of tying with the product release cycle

Author: I would like to leave that as another RFC.

Author: And it is probably a good idea to keep identical version with tidb, before it becomes a real "standalone" parser.

Reviewer 4: Yes, IMHO semver is a bigger topic than this, considering all components of TiDB & other components.

- What about the bloated parser.go file? Nobody wants it, and it increases the size of tidb repo.

> We can not simply remove it. Importing pingcap/parser to build applications needs parser.go file, but go lacks build scripts in rust or `postbuild` in nodejs to auto-generate parser.go.

Reviewer 3: ```suggestion
We can not simply remove it. Importing pingcap/parser to build applications needs parser.go file, but go module lacks build scripts in rust or `postbuild` in nodejs to auto-generate parser.go.
```

> We can not simply remove it. Importing pingcap/parser to build applications needs parser.go file, but go module lacks build scripts in rust or `postbuild` in nodejs to auto-generate parser.go.

Reviewer: ```suggestion
We can not simply remove it. Importing pingcap/parser to build applications needs parser.go file, but go module doesn't support something like build scripts in rust or `postbuild` in nodejs to auto-generate parser.go.
```
